### PR TITLE
PEP 768: Safe external debugger interface for CPython

### PR DESCRIPTION
**Pull request by pablogsal**

### CODE REVIEW

File: peps/pep-0778.rst [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+Python tools face this same challenge of code injection, but with an additional
+layer of complexity. Not only do they need to implement the above mechanism,
+they must also understand and safely interact with CPython's runtime state,
+including the interpreter loop, garbage collector, thread state, and reference
+counting system. This combination of low-level system manipulation and
+deep domain specific interpreter knowledge makes implementing Python debugging tools
+exceptionally difficult.
+
+The few tools ( see for example [1]_ and [2]_ )that do attempt this resort to
```
Comment (JelleZijlstra) [inline peps/pep-0778.rst:58]:
```suggestion
The few tools (see for example [1]_ and [2]_) that do attempt this resort to
```

File: peps/pep-0778.rst [hunk 2]
```diff
@@ -0,0 +1,359 @@
+PEP: 778
+Title: Safe external debugger interface for CPython
+Author: Pablo Galindo Salgado <pablogsal@python.org>, Matt Wozniski <godlygeek@gmail.com>, Ivona Stojanovic <stojanovic.i@hotmail.com>
+Sponsor: Pablo Galindo Salgado
```
Comment (JelleZijlstra) [inline peps/pep-0778.rst:4]: You can omit this since there's a core dev author
```suggestion
```

File: peps/pep-0778.rst [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+Rather than forcing tools to work around interpreter limitations with unsafe
+code injection, we can extend CPython with a proper debugging interface that
+guarantees safe execution. By adding a few thread state fields and integrating
+with the interpreter's existing evaluation loop, we can ensure debugging
+operations only occur at well-defined safe points. This eliminates the
+possibility of crashes and corruption while maintaining zero overhead during
+normal execution.
+
+
```
Comment (JelleZijlstra) [inline peps/pep-0778.rst:90]:
```suggestion
```
Too many blank lines

**Review by JelleZijlstra — APPROVED** [review]
This sounds great! Just a few nits.

File: peps/pep-0778.rst [hunk 4]
```diff
@@ -0,0 +1,357 @@
+PEP: 778
+Title: Safe external debugger interface for CPython
+Author: Pablo Galindo Salgado <pablogsal@python.org>, Matt Wozniski <godlygeek@gmail.com>, Ivona Stojanovic <stojanovic.i@hotmail.com>
+Discussions-To: xxxxx
+Status: Draft
+Type: Standards Track
+Topic: Interpreter Core
```
Comment (hugovk) [inline peps/pep-0778.rst:7]: Topics are these: https://peps.python.org/topic/

```suggestion
```

File: peps/pep-0778.rst [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+Debugging Python processes in production and live environments presents unique
+challenges. Developers often need to analyze application behavior without
+stopping or restarting services, which is especially crucial for
+high-availability systems. Common scenarios include diagnosing deadlocks,
+inspecting memory usage, or investigating unexpected behavior in real-time.
+
+Very few Python tools can attach to running processes, primarily because doing
+so requires deep expertise in both operating system debugging interfaces and
+CPython internals. While C/C++ debuggers like gdb and lldb can attach to
```
Comment (hugovk) [inline peps/pep-0778.rst:36]:
```suggestion
CPython internals. While C/C++ debuggers like GDB and LLDB can attach to
```

File: peps/pep-0778.rst [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+layer of complexity. Not only do they need to implement the above mechanism,
+they must also understand and safely interact with CPython's runtime state,
+including the interpreter loop, garbage collector, thread state, and reference
+counting system. This combination of low-level system manipulation and
+deep domain specific interpreter knowledge makes implementing Python debugging tools
+exceptionally difficult.
+
+The few tools (see for example [1]_ and [2]_) that do attempt this resort to
+suboptimal and unsafe methods,
+using system debuggers like gdb and lldb to forcefully inject code. This
```
Comment (hugovk) [inline peps/pep-0778.rst:59]:
```suggestion
using system debuggers like GDB and LLDB to forcefully inject code. This
```

File: peps/pep-0778.rst [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+  consistent security semantics across platforms.
+
+All mechanisms ensure that:
+
+1. Only authorized processes can read/write memory
+2. The same security model that governs traditional debugger attachment applies
+3. No additional attack surface is exposed beyond what the OS already provides for debugging
+
+The memory operations themselves are well-established and have been used safely
+for decades in tools like gdb, lldb, and various system profilers.
```
Comment (hugovk) [inline peps/pep-0778.rst:308]:
```suggestion
for decades in tools like GDB, LLDB, and various system profilers.
```

File: peps/pep-0778.rst [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+stopping or restarting services, which is especially crucial for
+high-availability systems. Common scenarios include diagnosing deadlocks,
+inspecting memory usage, or investigating unexpected behavior in real-time.
+
+Very few Python tools can attach to running processes, primarily because doing
+so requires deep expertise in both operating system debugging interfaces and
+CPython internals. While C/C++ debuggers like gdb and lldb can attach to
+processes using well-understood techniques, Python tools must implement all of
+these low-level mechanisms plus handle additional complexity. For example, when
+gdb needs to execute code in a target process, it:
```
Comment (hugovk) [inline peps/pep-0778.rst:39]:
```suggestion
GDB needs to execute code in a target process, it:
```

File: peps/pep-0778.rst [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+Python tools face this same challenge of code injection, but with an additional
+layer of complexity. Not only do they need to implement the above mechanism,
+they must also understand and safely interact with CPython's runtime state,
+including the interpreter loop, garbage collector, thread state, and reference
+counting system. This combination of low-level system manipulation and
+deep domain specific interpreter knowledge makes implementing Python debugging tools
+exceptionally difficult.
+
+The few tools (see for example [1]_ and [2]_) that do attempt this resort to
+suboptimal and unsafe methods,
```
Comment (hugovk) [inline peps/pep-0778.rst:58]: The footnotes are just links and no extra info, so let's save our readers a click and link directly to them:
```suggestion
The few tools (see for example `DebugPy
<https://github.com/microsoft/debugpy/blob/43f41029eabce338becbd1fa1a09727b3cfb1140/src/debugpy/_vendored/pydevd/pydevd_attach_to_process/linux_and_mac/attach.cpp#L4>`__
and `Memray
<https://github.com/bloomberg/memray/blob/main/src/memray/_memray/inject.cpp>`__)
that do attempt this resort to suboptimal and unsafe methods,
```

File: peps/pep-0778.rst [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+where they are in memory.
+
+Attachment Protocol
+-------------------
+When a debugger wants to attach to a Python process, it follows these steps:
+
+1. Locate ``PyRuntime`` structure in the process:
+   - Find Python binary (executable or libpython) in process memory (OS dependent process)
+   - Extract ``.PyRuntime`` section offset from binary's format (ELF/Mach-O/PE)
+   - Calculate the actual ``PyRuntime`` address in the running process by relocating the offset to the binary's load address
```
Comment (hugovk) [inline peps/pep-0778.rst:182]: I think we need a newline here for the sublist to render as a list and not a paragraph:
```suggestion
1. Locate ``PyRuntime`` structure in the process:

   - Find Python binary (executable or libpython) in process memory (OS dependent process)
   - Extract ``.PyRuntime`` section offset from binary's format (ELF/Mach-O/PE)
   - Calculate the actual ``PyRuntime`` address in the running process by relocating the offset to the binary's load address
```

File: peps/pep-0778.rst [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+2. Access debug offset information by reading the ``_Py_DebugOffsets`` at the start of the ``PyRuntime`` structure.
+
+3. Use the offsets to locate the desired thread state
+
+4. Use the offsets to locate the debugger interface fields within that thread state
+
+5. Write control information:
+   - Write python code to be executed into the ``debugger_script`` field in ``_PyRemoteDebuggerSupport``
+   - Set ``debugger_pending_call`` flag in ``_PyRemoteDebuggerSupport``
+   - Set ``_PY_EVAL_PLEASE_STOP_BIT`` in the ``eval_breaker`` field
```
Comment (hugovk) [inline peps/pep-0778.rst:193]:
```suggestion
5. Write control information:

   - Write python code to be executed into the ``debugger_script`` field in ``_PyRemoteDebuggerSupport``
   - Set ``debugger_pending_call`` flag in ``_PyRemoteDebuggerSupport``
   - Set ``_PY_EVAL_PLEASE_STOP_BIT`` in the ``eval_breaker`` field
```

**Review by hugovk — COMMENTED** [review]
Please add yourself to CODEOWNERS.

File: peps/pep-0778.rst [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+
+References
+==========
+
+.. [1] DebugPy
+   https://github.com/microsoft/debugpy/blob/43f41029eabce338becbd1fa1a09727b3cfb1140/src/debugpy/_vendored/pydevd/pydevd_attach_to_process/linux_and_mac/attach.cpp#L4
+
+.. [2] Memray
+   https://github.com/bloomberg/memray/blob/main/src/memray/_memray/inject.cpp
+
```
Comment (hugovk) [inline peps/pep-0778.rst:352]:
```suggestion
```

File: peps/pep-0778.rst [hunk 13]
```diff
@@ -0,0 +1,352 @@
+PEP: 778
```
Comment (hugovk) [inline peps/pep-0778.rst:1]: 778 is already assigned for "Supporting Symlinks in Wheels": https://github.com/python/peps/pull/3786

768 is the next available in sequence:
```console
❯ pip install pepotron --quiet && pep next
Next available PEP: 768
```

```suggestion
PEP: 768
```

**Comment by pablogsal** [comment]
Thanks a lot for your fantastic comments as always @JelleZijlstra @hugovk

**Pull request merged**